api/localizer/writer.go

```diff
@@ -0,0 +1,114 @@
+var errAbsPath = errors.New("absolute paths not yet supported")
+
+func makeAndClean(path string, fSys filesys.FileSystem) (filesys.ConfirmedDir, error) {
+	var cleanPath filesys.ConfirmedDir
+	var err error
+
+	if err = fSys.MkdirAll(path); err != nil {
+		return "", errors.Wrapf(err, filesys.ErrMkDirAll.Error())
+	}
+	if cleanPath, _, err = fSys.CleanedAbs(path); err != nil {
+		return "", errors.Wrapf(err, filesys.ErrCleanAbs.Error())
+	}
+
+	return cleanPath, nil
+}
+
+// NewWriter is the factory method for Writer.
+func NewWriter(target filesys.ConfirmedDir, scope filesys.ConfirmedDir, newDir filesys.ConfirmedDir, fSys filesys.FileSystem) (*Writer, error) {
+	inNewDir, _ := filepath.Rel(scope.String(), target.String())
+	dst, dstErr := makeAndClean(newDir.Join(inNewDir), fSys)
+	if dstErr != nil {
+		return nil, errors.Wrapf(dstErr, "unable to write to newDir")
+	}
+
+	return &Writer{
+		fSys:   fSys,
+		newDir: newDir,
+		root:   target,
+		dstDir: dst,
+	}, nil
+}
+
+// Writer takes care of writing localized files to their appropriate destination
+// in newDir for command kustomize localize. Each Writer is associated with a
+// kustomization root, either remote or local, referenced by target and writes
+// files that the root directly references.
+//
+// Writer is meant to work with loader, which performs the path checks.
+type Writer struct {
+	fSys filesys.FileSystem
+	// newDir argument, as defined by kustomize localize
+	newDir filesys.ConfirmedDir
+	// path to local copy of kustomization root for which Writer writes referenced
+	// files
+	root filesys.ConfirmedDir
+	// directory in newDir that mirrors root
+	dstDir filesys.ConfirmedDir
+}
+
+func (wr *Writer) getLocalizePath(u *url.URL) string {
+	pathParts := strings.Split(strings.Trim(u.Path, urlSeparator), urlSeparator)
+	return filepath.Join(localizeDir, u.Hostname(), filepath.Join(pathParts...))
+}
+
+// Write writes content to the newDir location that corresponds to path,
+// relative to the kustomization root that this Writer is associated with.
+// Write returns a path localized to the newly written file.
+//
+// Path must be a valid file path following kustomize rules. For instance,
+// Load(path) should run without error. Write only throws errors associated with
+// writing content.
+func (wr *Writer) Write(path string, content []byte) (string, error) {
+	var localizedPath string
+	switch {
+	case loader.HasRemoteFileScheme(path):
+		u, _ := url.Parse(path)
+		// version always present in raw github urls
+		localizedPath = wr.getLocalizePath(u)
+	case !filepath.IsAbs(path):
+		// avoid symlinks; only write file corresponding to actual location in root
+		// avoid path that Load() shows to be in root, but may traverse outside
+		// temporarily; for example, ../root/config; problematic for rename and
+		// relocation
+		locDir, locFile, _ := wr.fSys.CleanedAbs(wr.root.Join(path))
+		if locDir.HasPrefix(wr.newDir) { // newDir can be inside scope or target
```

Comment:
Is it always the case that `wr.newDir` would be the prefix? Is there any possible way it could have something like `some/other/filepath/newDir/....` ?

Reply:
> Is it always the case that `wr.newDir` would be the prefix? Is there any possible way it could have something like `some/other/filepath/newDir/....` ?

Yes, as far as I know, `wr.newDir` will always be the prefix if the file is in the destination directory for `kustomize localize` because `locDir` and `wr.newDir` are both [`filesys.ConfirmedDir`](https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/filesys/confirmeddir.go#L12)s. `ConfirmedDir`s are straightened absolute paths devoid of symlinks. Because a `ConfirmedDir` path must traverse in 1 direction, aka cannot enter and exist the same directory, a `ConfirmedDir` cannot diverge from `wr.newDir`, but still point into `wr.newDir`.

That said, `locDir.Join(locFile)` can point into `wr.newDir`, in the presence of a hard link, without `wr.newDir` prefixing `locDir`, but because `wr.newDir` did not exist before the current `kustomize localize` call, this is impossible.

Reply:
Got it, thanks!